openedx/openedx-platform
# CMS API serializers and validations

Internal ticket: https://2u-internal.atlassian.net/browse/TNL-11077

This adds validations to all new CMS API endpoints via serializers.

Validations follow the "strong parameter" concept:
https://www.mintbit.com/blog/securing-ruby-on-rails-applications-part-3-use-strong-parameters#:~:text=Strong%20parameters%20are%20a%20feature,parameters%20to%20a%20controller%20action.

How to test:
- Log into local studio as course author (preferably superadmin)
- go to http://localhost:18010/cms-api/ui/
- open the tabs of the different POST and PUT endpoints
- the body should show all important fields necessary for demoing the CMS API
- transcripts, file assets, and videos/images should have multipart/form-data, others application/json preselected as content type in the dropdown
- Unzip and Import the JSON file linked below into insomnia
- test out the different POST requests by replacing course_key and block_id and other stuff in the paths to work with your own courses
- They should return 200 or 201 responses when sent with correct urls and valid example data (e.g. select a file for upload from your own drive)
- Now add extra fields to the requests that are not in the schema, for example "foo": "bar" - response should be 400
- Change the field value type of a field, for example from string to boolean (e.g. in POST xblock: replace "category": "vertical" with "category": true) - response should be 400
- Now test the demo reference app application. Make sure it points to your local edx-platform urls. It should still work, there should be no failed requests or failed validations. -- this depends on [this reference app PR ](https://github.com/2uinc/xblock-api-experiments/pull/21/files).

[Insomnia_validations_test.json.zip](https://github.com/openedx/edx-platform/files/12689192/Insomnia_validations_test.json.zip)
